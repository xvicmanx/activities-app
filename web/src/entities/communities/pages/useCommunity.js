// @flow

import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchCommunity } from '../redux/CommunitiesActions';
import { readToken } from '../../users/redux/UsersActions';

const useCommunity = (id: number | string): ?Object => {
  const { Communities } = useSelector((state) => state);
  const dispatch = useDispatch();
  const token = readToken();

  useEffect(() => {
    dispatch(fetchCommunity(token, id));
  }, [dispatch, token, id]);

  const { community } = Communities;

  return community;
};

export default useCommunity;
