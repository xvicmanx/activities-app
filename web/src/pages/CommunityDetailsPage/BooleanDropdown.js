import React, { useEffect } from 'react';
import { Heading } from 'react-bulma-components';
import { useSelector, useDispatch } from 'react-redux';

import { fetchUsers } from '../../redux/Users/UsersActions';

import {
  DropDown,
  LoadingIndicator,
} from '../../components';
import { readTokenFromCookie } from '../../redux/Users/UsersActions';

const BooleanDropdown = ({ name, value, onChange }) => (
  <DropDown
    name={name}
    items={
      [
        {
          id: true,
          name: 'Si',
        },
        {
          id: false,
          name: 'No',
        },
      ]
    }
    onChange={onChange}
    placeholder="Seleccione"
    value={value}
  />
);

export default BooleanDropdown;
