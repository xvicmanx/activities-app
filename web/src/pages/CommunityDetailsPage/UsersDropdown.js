import React, { useEffect } from 'react';
import { Heading } from 'react-bulma-components';
import { useSelector, useDispatch } from 'react-redux';

import { fetchUsers } from '../../redux/Users/UsersActions';

import {
  DropDown,
  LoadingIndicator,
} from '../../components';
import { readTokenFromCookie } from '../../redux/Users/UsersActions';

const UsersDropdown = ({ name, value, onChange }) => {
  const { Users } = useSelector(state => state);
  const dispatch = useDispatch();
  const token = readTokenFromCookie();

  useEffect(() => {
    dispatch(fetchUsers(token));
  }, [dispatch]);

  if (Users.isLoading) {
    return <LoadingIndicator />;
  }

  return (
    <DropDown
      name={name}
      items={Users.users || []}
      onChange={onChange}
      placeholder="Seleccione usuario"
      value={value}
    />
  );
};

export default UsersDropdown;
