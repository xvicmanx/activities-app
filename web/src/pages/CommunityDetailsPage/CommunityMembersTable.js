import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import CRUDTable, { CreateForm, DeleteForm, Fields, Field } from 'react-crud-table';

import { readTokenFromCookie } from '../../redux/Users/UsersActions';
import { addMember } from '../../redux/Communities/CommunitiesActions';
import UsersDropdown from './UsersDropdown';
import BooleanDropdown from './BooleanDropdown';

const styles = {
  container: {
    margin: "auto",
    width: "fit-content",
  },
};

const MembersSelectRenderer = ({ field }) => (
  <UsersDropdown {...field}/>
);

const CoordinatesSelectRenderer = ({ field }) => (
  <BooleanDropdown {...field} />
);


const CommunityMembersTable = ({ communityId, members }) => {
  const dispatch = useDispatch();
  const token = readTokenFromCookie();
  return (
    <CRUDTable
      caption="Miembros"
      items={members}
    >
      <Fields>
        <Field
          name="memberId"
          label="Miembro"
          hideFromTable
          render={MembersSelectRenderer}
        />
        <Field
          name="name"
          label="Nombre"
          placeholder="Nombre"
          hideInCreateForm
        />
        <Field
          name="coordinates"
          label="Coordinador?"
          tableValueResolver={item => item.coordinates ? 'Si' : 'No'}
          render={CoordinatesSelectRenderer}
        />
      </Fields>
      <CreateForm
        title="Agregar miembro"
        message="Agregar miembro"
        trigger="Agregar miembro"
        onSubmit={item => {
          dispatch(addMember(
            token,
            communityId,
            item.memberId,
            item.coordinates,
          ));
          return Promise.resolve();
        }}
        submitText="Agregar"
        validate={(values) => {
          const errors = {};
          
          if (!values.memberId) {
            errors.name = 'Por favor, seleccione miembro';
          }
  
          return errors;
        }}
      />
    </CRUDTable>
  );
};

export default CommunityMembersTable;
