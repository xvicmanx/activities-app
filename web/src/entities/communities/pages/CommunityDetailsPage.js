import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { faBuilding } from '@fortawesome/free-solid-svg-icons';
import { Breadcrumb } from 'react-bulma-components';
import { Link, useParams } from 'react-router-dom';

import { fetchCommunity } from '../redux/CommunitiesActions';
import { readTokenFromCookie } from '../../users/redux/UsersActions';
import Page from '../../../components/Page/Page';
import { LoadingIndicator } from '../../../components';
import CommunityMembersTable from '../components/CommunityMembersTable';


const CommunityDetailsPage = () => {
  const { id } = useParams();
  const { Communities } = useSelector(state => state);
  const dispatch = useDispatch();
  const token = readTokenFromCookie();

  useEffect(() => {
    dispatch(fetchCommunity(
      token,
      id,
    ));
  }, [dispatch, token]);

  const { community } = Communities;

  if (!community) {
    return <LoadingIndicator />;
  }

  return (
    <Page
      title={`Detalles de comunidad: ${community.name}`}
      icon={faBuilding}
    >
      <div>
        <Breadcrumb
          renderAs={Link}
          hrefAttr="to"
          items={[
            {
              name: 'Comunidades',
              url: '/communities',
            }, {
              name: community.name,
              url: `/communities/${community.id}`,
              active: true,
            },
          ]}
        />
        <CommunityMembersTable communityId={community.id} members={community.members} />
      </div>
    </Page>
  );
};

export default CommunityDetailsPage;
