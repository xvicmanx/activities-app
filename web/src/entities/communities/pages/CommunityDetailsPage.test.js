import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import CommunityDetailsPage from './CommunityDetailsPage';

jest.mock('./useCommunity');
jest.mock('react-router-dom', () => ({
  useParams: () => ({ id: 1 }),
  Link: () => 'Link',
}));

import useCommunity from './useCommunity';

describe('CommunityDetailsPage', () => {
  beforeEach(() => {
    useCommunity.mockClear();
  });

  it('renders properly when community available', () => {
    const community = {
      id: 1,
      name: 'Test community',
      members: [{
        id: 2,
        name: 'John',
      }]
    };
    const renderer = new ShallowRenderer();
    
    useCommunity.mockImplementation(() => community);

    const result = renderer.render(<CommunityDetailsPage />);

    expect(useCommunity).toHaveBeenCalledTimes(1);
    expect(useCommunity).toHaveBeenCalledWith(1);
    expect(result).toMatchSnapshot();
  });

  it('renders properly when community not available', () => {
    const renderer = new ShallowRenderer();
    
    useCommunity.mockImplementation(() => null);

    const result = renderer.render(<CommunityDetailsPage />);

    expect(useCommunity).toHaveBeenCalledTimes(1);
    expect(useCommunity).toHaveBeenCalledWith(1);
    expect(result).toMatchSnapshot();
  });
});
