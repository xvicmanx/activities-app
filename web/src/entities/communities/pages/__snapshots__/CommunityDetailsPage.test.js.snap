// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunityDetailsPage renders properly when community available 1`] = `
<Page
  icon={
    Object {
      "icon": Array [
        448,
        512,
        Array [],
        "f1ad",
        "M436 480h-20V24c0-13.255-10.745-24-24-24H56C42.745 0 32 10.745 32 24v456H12c-6.627 0-12 5.373-12 12v20h448v-20c0-6.627-5.373-12-12-12zM128 76c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40c0 6.627-5.373 12-12 12h-40c-6.627 0-12-5.373-12-12V76zm0 96c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40c0 6.627-5.373 12-12 12h-40c-6.627 0-12-5.373-12-12v-40zm52 148h-40c-6.627 0-12-5.373-12-12v-40c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40c0 6.627-5.373 12-12 12zm76 160h-64v-84c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v84zm64-172c0 6.627-5.373 12-12 12h-40c-6.627 0-12-5.373-12-12v-40c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40zm0-96c0 6.627-5.373 12-12 12h-40c-6.627 0-12-5.373-12-12v-40c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40zm0-96c0 6.627-5.373 12-12 12h-40c-6.627 0-12-5.373-12-12V76c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v40z",
      ],
      "iconName": "building",
      "prefix": "fas",
    }
  }
  title="Detalles de comunidad: Test community"
>
  <div>
    <Breadcrumb
      hrefAttr="to"
      items={
        Array [
          Object {
            "name": "Comunidades",
            "url": "/communities",
          },
          Object {
            "active": true,
            "name": "Test community",
            "url": "/communities/1",
          },
        ]
      }
      renderAs={[Function]}
    />
    <CommunityMembersTable
      communityId={1}
      members={
        Array [
          Object {
            "id": 2,
            "name": "John",
          },
        ]
      }
    />
  </div>
</Page>
`;

exports[`CommunityDetailsPage renders properly when community not available 1`] = `<LoadingIndicator />`;
